packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -833,34 +836,35 @@ class _TextSelectionHandleOverlayState
     //
     // For the end handle we compute the rectangles that encompass the range
     // of the last full selected grapheme cluster at the end of the selection.
+    //
+    // Only calculate start/end handle rects if the text in the previous frame
+    // is the same as the text in the current frame. This is done because
+    // widget.renderObject contains the renderEditable from the previous frame.
+    // If the text changed between the current and previous frames then
+    // widget.renderObject.getRectForComposingRange might fail. In cases where
+    // the current frame is different from the previous we fall back to
+    // widget.renderObject.preferredLineHeight.
     final InlineSpan span = widget.renderObject.text!;
-    final String text = span.toPlainText();
+    final String prevText = span.toPlainText();
+    final String currText = widget.selectionDelegate.textEditingValue.text;
     final int firstSelectedGraphemeExtent;
     final int lastSelectedGraphemeExtent;
-    final TextSelection? selection = widget.renderObject.selection;
+    final TextSelection selection = widget.selection;
+    final Rect? startHandleRect;
+    final Rect? endHandleRect;
 
-    if (selection != null && selection.isValid && !selection.isCollapsed) {
-      final String selectedGraphemes = selection.textInside(text);
+    if (prevText == currText && selection != null && selection.isValid && !selection.isCollapsed) {
+      final String selectedGraphemes = selection.textInside(currText);
       firstSelectedGraphemeExtent = selectedGraphemes.characters.first.length;
       lastSelectedGraphemeExtent = selectedGraphemes.characters.last.length;
       assert(firstSelectedGraphemeExtent <= selectedGraphemes.length && lastSelectedGraphemeExtent <= selectedGraphemes.length);
+      startHandleRect = widget.renderObject.getRectForComposingRange(TextRange(start: selection.start, end: selection.start + firstSelectedGraphemeExtent));
+      endHandleRect = widget.renderObject.getRectForComposingRange(TextRange(start: selection.end - lastSelectedGraphemeExtent, end: selection.end));
     } else {
-      // The call to selectedGraphemes.characters.first/last will throw a state
-      // error if the given text is empty, so fall back to first/last character
-      // range in this case.
-      //
-      // The call to widget.selection.textInside(text) will return a RangeError
-      // for a collapsed selection, fall back to this case when that happens.
-      firstSelectedGraphemeExtent = 0;
-      lastSelectedGraphemeExtent = 0;
+      startHandleRect = null;
```

Comment:
This can be removed - startHandleRect/endHandleRect will be initialized to null by default

Reply:
Do you mean I should initialize them to null by default, or is this done automatically. I tried removing this `else` block but got some errors saying that I cannot use the rects before they have been initialized.

Reply:
The variables will be initialized to null by default if they are not `final`.  If they are `final`, then they must be explicitly assigned.

Either way is fine.